tools/kagami/src/crypto.rs

```diff
@@ -1,14 +1,207 @@
+    #[clap(long, group = "public_key")]
+    public_key_file: Option<PathBuf>,
+    /// Path to json-serialized keypair
+    #[clap(long, short, group = "private_key", group = "public_key")]
+    keypair_file: Option<PathBuf>,
+    /// Unique id of blockchain
+    #[clap(long)]
+    chain_id: ChainId,
+    /// Path to genesis json file
+    #[clap(long, short)]
+    genesis_file: PathBuf,
+    /// Output signed genesis config as a hex string
+    #[clap(long, default_value_t = true, group = "format")]
+    hex: bool,
+    /// Encode signed genesis block with SCALE (it is only supported with file output)
+    #[clap(long, short, default_value_t = false, group = "format")]
+    scale: bool,
+    /// Path to signed genesis output file (stdout by default)
+    #[clap(long, short)]
+    out_file: Option<PathBuf>,
+}
+
+#[derive(Debug)]
+enum KeyStorage<'a> {
+    FromFile(Vec<u8>),
+    FromCLI(&'a str),
+}
+
+fn get_key_raw<'a, P: AsRef<Path>>(
+    path: &Option<P>,
+    value: &'a Option<String>,
+) -> Result<KeyStorage<'a>, std::io::Error> {
+    match (path, value) {
+        (Some(path_buf), None) => Ok(KeyStorage::FromFile(fs::read(path_buf)?)),
+        (None, Some(hex)) => Ok(KeyStorage::FromCLI(hex.as_str())),
+        _ => unreachable!("Clap group invariant"),
+    }
+}
+
+fn read_keypair<P: AsRef<Path>>(path: P) -> Result<KeyPair> {
+    let bytes = fs::read(path)?;
+    Ok(serde_json::from_slice(bytes.as_slice())?)
+}
+
+impl SignGenesisArgs {
+    fn get_private_key(&self) -> Result<PrivateKey> {
+        let private_key_bytes = get_key_raw(&self.private_key_file, &self.private_key_string)?;
+        match private_key_bytes {
+            KeyStorage::FromFile(bytes) => {
+                PrivateKey::from_bytes(self.algorithm.0, bytes.as_slice()).wrap_err_with(|| {
+                    eyre!(
+                        "Failed to parse private key from bytes for algorithm `{}`",
+                        self.algorithm
+                    )
+                })
+            }
+            KeyStorage::FromCLI(hex) => {
+                PrivateKey::from_hex(self.algorithm.0, hex).wrap_err_with(|| {
+                    eyre!(
+                        "Failed to parse private key from hex for algorithm `{}`",
+                        self.algorithm
+                    )
+                })
+            }
+        }
+    }
+
+    fn get_public_key(&self) -> Result<PublicKey> {
+        let public_key_bytes = get_key_raw(&self.public_key_file, &self.public_key_string)?;
+        match public_key_bytes {
+            KeyStorage::FromFile(bytes) => {
+                PublicKey::from_bytes(self.algorithm.0, bytes.as_slice()).wrap_err_with(|| {
+                    eyre!(
+                        "Failed to parse public key from bytes for algorithm `{}`",
+                        self.algorithm
+                    )
+                })
+            }
+            KeyStorage::FromCLI(hex) => {
+                if hex.starts_with(ED25519_PREFIX) {
```

Comment:
`PublicKey` is always represented as a multihash string. Look at other places where `PublicKey` is parsed from external input and re-use that way, please.